lib/IDE/CompletionLookup.cpp

```diff
@@ -2969,6 +2988,44 @@ void CompletionLookup::getAttributeDeclParamCompletions(
       break;
     }
     break;
+  case CustomSyntaxAttributeKind::StorageRestrictions:
+    if (!HasLabel) {
+      bool suggestInitializesLabel = false;
+      bool suggestAccessesLabel = false;
+      bool suggestArgument = false;
+      switch (static_cast<StorageRestrictionsCompletionKind>(ParamIndex)) {
```

Comment:
The question is: How far do we want to add a separate parameter for it? I would say the two options are to either keep it as-is or to do all of the following
- Add a parameter to `getAttributeDeclParamCompletions`
- Add a member to `CodeCompletionCallbacksImpl`, 
- Add a new `completeStorageRestrictionsAttribute` function to `CodeCompletionCallbacksImpl` that takes a `StorageRestrictionsCompletionKind`

I was mostly held back by adding yet another member to `CodeCompletionCallbacksImpl` but if you prefer that, I’m happy to do it. @rintaro Do you have any opinions?

Reply:
Ah sorry I missed the fact that this is all being stored on the `CodeCompletionCallbacksImpl` 😅 I kind of wonder if we should break this out into a class hierarchy, similar to how we model AST nodes, e.g there should be a base class that stores the completion kind, and then subclasses have their own completion kind with their state, and then there can be further subclasses that store more state, and we can use `dyn_cast` and such when doing the completion lookup. That's a much bigger refactoring though, I'm happy to leave it as this for now.

Reply:
Good idea. I’ll draft a follow-up PR for it.

Reply:
OK, I just started doing it and we would need to generate ~300-400 lines of boilerplate just for those class definitions (30 completion kinds, each with a class, member, initializer and `classof` function). While I think it would be nice, it also doesn’t really fit into the current design of `CodeCompletionCallbacksImpl`, which relies a lot on the ad-hoc nature (eg. setting one of the members in a setter or type checking in `doneParsing` if `ParsedExpr` is not a `nullptr`).

We need to revisit how code completion works with ASTGen anyway, so I would prefer to not make any changes right now and instead re-design it in the light of ASTGen.

Reply:
Fair enough, sounds good!